From 99ab588bf06841969b6042aa98fcb2b6aa4ac0a6 Mon Sep 17 00:00:00 2001
From: Joakim Bech <joakim.bech@gmail.com>
Date: Sat, 23 May 2026 23:31:12 +0200
Subject: [PATCH] build: use lld for RISC-V cross-compilation

The host x86_64 ld.bfd does not support elf32lriscv. Clang must be
told to use lld, which ships with the LLVM toolchain and handles all
RISC-V targets.
---
 hw/application_fpga/Makefile      | 2 ++
 hw/application_fpga/apps/Makefile | 1 +
 2 files changed, 3 insertions(+)

diff --git a/hw/application_fpga/Makefile b/hw/application_fpga/Makefile
index d95acc2..4141169 100644
--- a/hw/application_fpga/Makefile
+++ b/hw/application_fpga/Makefile
@@ -219,11 +219,13 @@ tools/b2s/b2s:
 # Included in the bitstream.
 #-------------------------------------------------------------------
 LDFLAGS = \
+	-fuse-ld=lld \
 	-T $(P)/fw/tk1/firmware.lds \
 	-Wl,--cref,-M \
 	-L $(LIBDIR) -lcommon -lblake2s
 
 QEMU_LDFLAGS = \
+	-fuse-ld=lld \
 	-T $(P)/fw/tk1/qemu_firmware.lds \
 	-Wl,--cref,-M \
 	-L $(LIBDIR) -lcommon -lblake2s
diff --git a/hw/application_fpga/apps/Makefile b/hw/application_fpga/apps/Makefile
index 3c2bce7..7663493 100644
--- a/hw/application_fpga/apps/Makefile
+++ b/hw/application_fpga/apps/Makefile
@@ -40,6 +40,7 @@ ASFLAGS = \
 	-mno-relax
 
 LDFLAGS = \
+	-fuse-ld=lld \
 	-T $(LIBDIR)/app.lds \
 	-L $(LIBDIR) -lcrt0 -lcommon -lmonocypher -lblake2s
 
-- 
2.53.0

